typescript/radix-sdk/src/modules/query.ts

```diff
@@ -0,0 +1,769 @@
+      });
+
+    assert(
+      !(response.receipt as any).error_message,
+      `${(response.receipt as any).error_message}`,
+    );
+
+    const fee_summary: FeeSummary = (response.receipt as any).fee_summary;
+    const costing_parameters: CostingParameters = (response.receipt as any)
+      .costing_parameters;
+
+    const gasUnits =
+      BigInt(fee_summary.execution_cost_units_consumed) +
+      BigInt(fee_summary.finalization_cost_units_consumed);
+    const fee = BigInt(
+      new BigNumber(fee_summary.xrd_total_execution_cost)
+        .plus(BigNumber(fee_summary.xrd_total_finalization_cost))
+        .plus(BigNumber(fee_summary.xrd_total_storage_cost))
+        .times(new BigNumber(10).exponentiatedBy(18))
+        .toFixed(0),
+    );
+    const gasPrice =
+      (parseFloat(costing_parameters.execution_cost_unit_price) +
+        parseFloat(costing_parameters.finalization_cost_unit_price)) *
+      0.5; // average out the cost parameters to get a more accurate estimate
+
+    return {
+      gasUnits,
+      fee,
+      gasPrice,
+    };
+  }
+
+  public async getMetadata({ resource }: { resource: string }): Promise<{
+    name: string;
+    symbol: string;
+    description: string;
+    divisibility: number;
+  }> {
+    const details =
+      await this.gateway.state.getEntityDetailsVaultAggregated(resource);
+
+    const result = {
+      name:
+        (
+          details.metadata.items.find((i) => i.key === 'name')?.value
+            .typed as any
+        ).value ?? '',
+      symbol:
+        (
+          details.metadata.items.find((i) => i.key === 'symbol')?.value
+            .typed as any
+        ).value ?? '',
+      description:
+        (
+          details.metadata.items.find((i) => i.key === 'description')?.value
+            .typed as any
+        ).value ?? '',
+      divisibility: (details.details as any).divisibility as number,
+    };
+
+    return result;
+  }
+
+  public async getXrdMetadata(): Promise<{
+    name: string;
+    symbol: string;
+    description: string;
+    divisibility: number;
+  }> {
+    const xrdAddress = await this.getXrdAddress();
+    return this.getMetadata({ resource: xrdAddress });
+  }
+
+  public async getBalance({
+    address,
+    resource,
+  }: {
+    address: string;
+    resource: string;
+  }): Promise<bigint> {
+    const details =
+      await this.gateway.state.getEntityDetailsVaultAggregated(address);
+
+    const fungibleResource = details.fungible_resources.items.find(
+      (r) => r.resource_address === resource,
+    );
+
+    if (!fungibleResource) {
+      return BigInt(0);
+    }
+
+    if (fungibleResource.vaults.items.length !== 1) {
+      return BigInt(0);
+    }
+
+    const { divisibility } = await this.getMetadata({ resource });
+
+    return BigInt(
+      new BigNumber(fungibleResource.vaults.items[0].amount)
+        .times(new BigNumber(10).exponentiatedBy(divisibility))
+        .toFixed(0),
+    );
+  }
+
+  public async getXrdBalance({
```

Comment:
this class feels way too big. can we split it up somehow? for example, the xrd methods seem to inherit from some base method (getBalance in this case)